openforcefield/topology/molecule.py

```diff
@@ -1381,6 +1405,23 @@ def molecule_bond_index(self):
             raise ValueError("This Atom does not belong to a Molecule object")
         return self._molecule.bonds.index(self)
 
+    @property
+    def is_in_ring(self):
+        """
+        Return whether or not this atom is in a ring(s) (of any size)
+
+        """
+        if self._molecule is None:
+            raise NotAttachedToMoleculeError(
+                "This Bond does not belong to a Molecule object"
+            )
+
+        # Note that the definition here needs to be different than in Atom.is_in_ring,
+        # since a bond's two atoms may be in a ring, but _not the same_ ring;
+        # A MWE is biphenyl (see unit test)
+        # bonds_in_any_ring = [val[0] for val in self._molecule.chemical_environment_matches('[*;R*:1]~[*;R*:2]')]
+        return self.atom1.is_in_ring and self.atom2.is_in_ring
```

Comment:
A better approach (if this is something that chemical perception can do well) may be to just find all of the rings in a structure, listed them separately ring by ring, and iterate through them one at a time. This would get around the issue of a bond between atoms in different rings, but not in a ring itself.

Reply:
I think this is something that networkx might be able to do well and should be easy to implement as we already have the `to_networkx` method. I also think it might be more useful to know which ring an atom or bond is in rather than just if it is in a ring especially in the case of fragmenter.

Reply:
It's not 100% functional yet but I've gone down this path, which includes storing rings as a nested list in `Molecule._rings`. It's not really stress-tested nor is the performance evaluated, but it worked pretty well for some simple cases, like catching the non-ring bond in biphenyl. Thanks for the suggestion, and let me know if this helps your tools or if there are any useful test cases we should keep in mind here!

Reply:
Nice!

A possibility for a test here might be to compare the networkx cycle basis with RDKit RingInfo AtomRings  https://www.rdkit.org/docs/source/rdkit.Chem.rdchem.html#rdkit.Chem.rdchem.RingInfo

On the 1156 molecules in [the AlkEthOH rings set](https://github.com/openforcefield/open-forcefield-data/blob/master/Model-Systems/AlkEthOH_distrib/AlkEthOH_rings.smi), these two methods agreed with each other.

```python
def _canonicalize(rings):
    """remove atom-ordering and ring-ordering, for the purpose
    of checking whether two collections of rings describe the same collections
    of atoms
    
    >>> rings = [[5, 4, 3, 2, 1, 0], [8, 7, 6, 4, 3]]
    >>> _canonicalize(rings)
    [(0, 1, 2, 3, 4, 5), (3, 4, 6, 7, 8)]
    """
    return sorted([tuple(sorted(ring)) for ring in rings])

def check_nx_cycle_basis_agrees_with_rdkit_ringinfo(offmol):
    # enumerate rings using networkx
    nx_rings = nx.cycle_basis(offmol.to_networkx())
    
    # enumerate rings using rdkit
    rdmol = offmol.to_rdkit()
    ring_info = rdmol.GetRingInfo()
    rd_rings = ring_info.AtomRings()
    
    # check if equivalent up to ordering
    return _canonicalize(nx_rings) == _canonicalize(rd_rings)
```

Reply:
Thanks for the write-up! That's really encouraging that it actually works on a relevant test set (more than the first three molecules to pop in my head)

One edge case I have dealt with in the past that worries me here is [_highly_ conjugated ring networks](https://github.com/mosdef-hub/foyer/issues/170#issuecomment-394779512). I used a SMILES drawer to make a not-very-physical slab of seven rings attached together (`C1C2C(CCC1)CC5C3C2CCC7C3C4C(CCC6C4C5CCC6)CC7`), and can get it to fail

```
In [9]: rdinfo.AtomRings()
Out[9]:
((0, 1, 2, 3, 4, 5),
 (6, 7, 8, 9, 1, 2),
 (10, 11, 12, 13, 8, 9),
 (16, 17, 18, 19, 14, 15),
 (21, 22, 23, 18, 19, 20),
 (24, 25, 12, 13, 14, 15),
 (7, 8, 13, 14, 19, 20))

In [10]: mol.rings
Out[10]:
[[9, 8, 7, 6, 2, 1],
 [12, 25, 24, 15, 14, 13],
 [19, 18, 17, 16, 15, 14],
 [20, 21, 22, 23, 18, 17, 16, 15, 14, 13, 8, 7],
 [20, 19, 14, 13, 8, 7],
 [9, 10, 11, 12, 13, 8],
 [5, 4, 3, 2, 1, 0]]
```

Networkx's [criteria](https://networkx.github.io/documentation/networkx-1.10/reference/generated/networkx.algorithms.cycles.cycle_basis.html) for the solution is ever so slightly different than how we would define rings, and there is a [different](https://github.com/mosdef-hub/foyer/blob/12e9e6d567be321a9d977c335c9f6422af04887b/foyer/smarts_graph.py#L244) implementation of the same functionality out there to deal with that. It's really unlikely to be an error that causes problems in the data we're working with now, but there's no way to predict when this discrepancy might pop up, so I had planned to switch it over to the longer approach.

However, this is giving me pause to consider which path is the better solution here, given that RDKit seems to have done the heavy lifting already.

Reply:
Another data point that would be useful here is whether or not we want these objects to generally be aware of their own rings (something that's implied by the `self._rings` caching), i.e. if that functionality is useful outside of the scope of the feature request here. If so, there could be a `ToolkitWrapper` function that implements this (presumably there is an OpenEye equivalent to this RDKit function). But if not, we don't need to go about implementing several different things on the backend for this purpose.

Reply:
It looks like the `nx.minimum_cycle_basis` function would help with this specific molecule and still works in general as it passes the test above from @maxentile. I think having the rings property on the molecule is really useful and I can imagine this being used in various areas outside of fragmenter. For example, this method will probably be used to check if dihedrals are in rings when submitting torsiondrives to qcarchive.

Reply:
Thanks for finding that - it seems to work perfectly and  I've implemented that and `Molecule.rings`/`Molecule.n_rings` with some basic tests

It's out of scope here, but it might be interesting to compare networkx, RDKit, and OpenEye methods for a larger data set since comparing methods wouldn't require labeling numbers of rings by hand

Reply:
>However, this is giving me pause to consider which path is the better solution here, given that RDKit seems to have done the heavy lifting already.

I did a quick random check to see if there are molecules in the [Enamine Diverse REAL drug-like set](https://enamine.net/library-synthesis/real-compounds/real-compound-libraries) where RDKit ring info disagrees with nx.minimum_cycle_basis. Of ~3000 random molecules checked, ~400 of them appear to have a discrepancy between nx.minimum_cycle_basis and RDKit ring info.

I haven't inspected these yet to say anything about how relevant the chemistry is, or whether RDKit vs. Networkx is giving the desired result, but in case this is useful for testing, here are the smiles encountered so far that cause discrepancies : https://gist.github.com/maxentile/64558c9e64b3394191cdd92594de760b

Reply:
Thanks for that I pulled out some molecules randomly and they seem to have a group like the molecule here and the networkx method seems to just miss one of the possible rings in the molecule. 
![Screenshot 2020-09-22 at 17 49 37](https://user-images.githubusercontent.com/30259414/93912706-012e6f80-fcfc-11ea-9055-8c356ddf0989.png)

Reply:
Looks like a common feature in that data set is that RDKit finds more rings:

```python3
In [11]: def n_rdkit(mol):
    ...:     rdmol = mol.to_rdkit()
    ...:     info = rdmol.GetRingInfo()
    ...:     rings = info.AtomRings()
    ...:     return len(rings)
    ...:

In [12]: for mol in mols:  # mols is from Fass's gist above
    ...:     assert n_rdkit(mol) > mol.n_rings  # using the networkx implementation
    ...:     # no exceptions raised

```

Pulling out a random molecule, I see a similar moiety to what @JoshHorton points out (not sure what its proper name is):

![image](https://user-images.githubusercontent.com/7935382/93932181-662a9b00-fce5-11ea-85dc-cfe804e4684d.png)


so on this basis I think we should trust RDKit's implementation and avoid networkx's out-of-the-box solutions.

Reply:
Well, this has certainly led down a deep rabbit hole, one that I didn't intend to go but should still be worthwhile for the toolkit's functionality.

* The world of graph theory does not seem to define any nouns that are close enough to what a chemist thinks of as a ring, so networkx's convenient functions will unfortunately not work. See above for bicyclics as a case that demonstrates the issue, norbornane being a simple starting point.
* RDKit seems to have done this already, see @maxentile's code above. An RDKit-based solution, I think should be implemented as a `ToolkitWrapper` function, which runs into an issue below.
* Foyer's solution to the problem (which takes only a networkx graph [1] and a setting for the maximum size of a ring it searches for) pretty closely matches RDKit's, but fails on some edge cases with multi-cyclics [2].
* OpenEye, for that matter, seems to [intentionally](https://docs.eyesopen.com/toolkits/python/oechemtk/ring.html#smallest-set-of-smallest-rings-sssr-considered-harmful) not have this functionality. It looks like it can be backed out with some hacks, but this is not a path I'm eager to go down.

I think the two solutions here are
1. Work toward patching the function that foyer uses to behave better in some test dataset of a few thousand molecules.
2. Write a `ToolkitWrapper` function that uses RDKit, and RDKit alone, to determine the rings in a molecule. This would be an issue if, for some reason, a user did not or could not use RDKit.



--
1. As an aside, I think it's really cool that much of the community uses common data structures like networkx graphs and NumPy arrays. I was able to conda install a package, import a function from it and just send it `mol.to_networkx()` out of the box, and it worked.

2. With the SMILES `'[H]C1(C2(C(C1(C(C2([H])[H])([H])SC3(C(OC3([H])[H])([H])[H])[H])[H])([H])[H])[H])[H]'` foyer finds a 10-membered ring by doubling back on itself while traversing the bicyclic group on the left. This may be patchable, I'm not sure without putting my thinking cap on. Moreso, this may be a case of the nasty complexity hinted at in the OpenEye docs linked above.

![image](https://user-images.githubusercontent.com/7935382/93937304-249dee00-fced-11ea-9c8a-2d65c82e1171.png)